rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/dap/variable/VariableSubfieldsVisitor.java

```diff
@@ -0,0 +1,188 @@
+        List<RascalVariable> result = new ArrayList<>();
+        Type type = visitedType.isTuple() ? visitedType : o.getType();
+        int max = count == -1 ? o.arity() : Math.min(o.arity(), startIndex+count);
+        for (int i = startIndex; i < max; i++) {
+            RascalVariable newVar = new RascalVariable(type.getFieldType(i), type.hasFieldNames() ? type.getFieldName(i) : Integer.toString(i), o.get(i));
+            addVariableToResult(newVar, result);
+        }
+        return result;
+    }
+
+    @Override
+    public List<RascalVariable> visitNode(INode o) throws RuntimeException {
+        List<RascalVariable> result = new ArrayList<>();
+
+        int max = count == -1 ? o.arity() : Math.min(o.arity(), startIndex+count);
+        for (int i = startIndex; i < max; i++) {
+            RascalVariable newVar = new RascalVariable(o.get(i).getType(), Integer.toString(i), o.get(i));
+            addVariableToResult(newVar, result);
+        }
+        return result;
+    }
+
+    @Override
+    public List<RascalVariable> visitConstructor(IConstructor o) throws RuntimeException {
+        List<RascalVariable> result = new ArrayList<>();
+
+        if(startIndex<o.arity()) {
+            int max = count == -1 ? o.arity() : Math.min(o.arity(), startIndex+count);
+            for (int i = startIndex; i < max; i++) {
+                RascalVariable newVar = new RascalVariable(o.getConstructorType().getFieldType(i), o.getConstructorType().hasFieldNames() ? o.getConstructorType().getFieldName(i) : Integer.toString(i), o.get(i));
+                addVariableToResult(newVar, result);
+            }
+        }
+
+        int remaining = count == -1 ? 1 : (startIndex+count) - o.arity();
+        if (o.mayHaveKeywordParameters() && remaining > 0) {
+            Map<String, IValue> parameters = o.asWithKeywordParameters().getParameters();
+            int startParametersIndex = startIndex<o.arity() ? 0 : startIndex-o.arity();
+            Object[] keys = parameters.keySet().toArray();
+            int max = count == -1 ? parameters.size() : Math.min(parameters.size(), startParametersIndex+remaining);
+            for(int i = startParametersIndex; i< max; i++){
+                String name = (String) keys[i];
+                IValue value = parameters.get(name);
+                RascalVariable newVar = new RascalVariable(value.getType(), '['+name+']', value);
+                addVariableToResult(newVar, result);
+            }
+        }
+
+        return result;
+    }
+
+    @Override
+    public List<RascalVariable> visitInteger(IInteger o) throws RuntimeException {
+        return new ArrayList<>();
+    }
+
+    @Override
+    public List<RascalVariable> visitMap(IMap o) throws RuntimeException {
+        List<RascalVariable> result = new ArrayList<>();
+
+        Object[] pairs = o.stream().toArray();
+        int max = count == -1 ? o.size() : Math.min(o.size(), startIndex+count);
+        for(int i = startIndex; i< max; i++){
```

Comment:
can this be written as an iterator over the map? instead of having to take a copy of the whole map? is there a limit in the API of `IMap`?

Reply:
I'd use this one:

```
/**
     * @return an iterator over the keys-value pairs of the map
     */
    public Iterator<Entry<IValue, IValue>> entryIterator();
```

Typically:
```
for (Entry<IValue,IValue> e : () -> myMap.entryIterator()) {
    IString s = (IString) e.getKey();
    ...
 }
 ```


Or: simply the `iterator()` method, IMap is also iterable, so you can: `for (IValue key : map)`.

Reply:
The lambda trick turns an iterator into an Iterable, such that the `for` loop knows what to do.